File: Astro-TFG/public/scripts/4.1-calculos-matematicos/prueba-3/version-1/javascript/main.js
function calculatePi(digits) {
    let pi = 0.0;
    for (let k = 0; k < digits; k++) {
        pi += (1 / Math.pow(16, k)) * (
            (4 / (8 * k + 1)) -
            (2 / (8 * k + 4)) -
            (1 / (8 * k + 5)) -
            (1 / (8 * k + 6))
        );
    }
    return pi;
}

function n_digits_pi(repetitions, digits) {
    let totalTime = 0;
    let totalMemory = 0;

    const startTotal = performance.now();

    for (let i = 0; i < repetitions; i++) {
        const start = performance.now();

        const piValue = calculatePi(digits);

        const end = performance.now();
        const memoryUsage = performance.memory ? (performance.memory.usedJSHeapSize / (1024 * 1024)) : 0;

        totalTime += (end - start);
        totalMemory += memoryUsage;
    }

    const endTotal = performance.now();
    const totalExecTime = (endTotal - startTotal).toFixed(2);

    const avgTime = Number((totalTime / repetitions).toFixed(2));
    const avgMemory = Number((totalMemory / repetitions).toFixed(2));

    const outputDiv = document.getElementById("javascript-output");

    // ET (Execution Time)

    let timeDiv = document.createElement("div");
    timeDiv.textContent = `Total ET: ${totalExecTime} ms`;
    outputDiv.appendChild(timeDiv);

    let avgTimeDiv = document.createElement("div");
    avgTimeDiv.textContent = `ET (avg, 1000x): ${avgTime} ms`;
    outputDiv.appendChild(avgTimeDiv);

    // RAM

    let avgMemoryDiv = document.createElement("div");
    avgMemoryDiv.textContent = `RAM (avg, 1000x): ${avgMemory} MB`;
    outputDiv.appendChild(avgMemoryDiv);
}

window.runJSBenchmark = function () {
    clearCell("javascript-output");
    window.showExecutionLoader();
    requestAnimationFrame(() => {
        setTimeout(() => {
            n_digits_pi(1_000, 1_000);
            window.hideExecutionLoader();
        }, 0);
    });
}

function clearCell(elementId) {
    document.getElementById(elementId).innerHTML = "";
}

File: Astro-TFG/public/scripts/4.1-calculos-matematicos/prueba-3/version-1/python/main.py
import time
import tracemalloc
import gc
import js  # type: ignore
from pyscript import display  # type: ignore


def calculate_pi(digits):
    pi = 0.0
    for k in range(digits):
        pi += (1 / (16**k)) * (
            (4 / (8 * k + 1)) -
            (2 / (8 * k + 4)) -
            (1 / (8 * k + 5)) -
            (1 / (8 * k + 6))
        )
    return pi


def n_digits_pi(repetitions, digits):
    total_time = 0
    total_memory = 0

    start_total = time.time()

    for _ in range(repetitions):
        tracemalloc.start()

        start = time.time()
        gc.collect()
        pi_value = calculate_pi(digits)

        end = time.time()
        memory_usage = tracemalloc.get_traced_memory()[1] / (1024 * 1024)  # MB
        tracemalloc.stop()

        total_time += (end - start) * 1000
        total_memory += memory_usage

    # ET (Execution Time)

    end_total = time.time()
    total_exec_time = round((end_total - start_total) * 1000, 2)
    avg_time = round(total_time / repetitions, 2)

    # RAM

    avg_memory = round(total_memory / repetitions, 2)

    display(f"Total ET (1000x): {total_exec_time} ms",
            target="pyscript-output")
    display(f"ET (avg, 1000x): {avg_time} ms", target="pyscript-output")
    display(f"RAM (avg, 1000x): {avg_memory} MB", target="pyscript-output")
    js.endTimerWebAssembly()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    n_digits_pi(1_000, 1_000)


js.window.run_py_benchmark = run_py_benchmark
